plugins/arrow-flight-rpc/src/main/java/org/opensearch/arrow/flight/bootstrap/FlightClientManager.java

```diff
@@ -0,0 +1,265 @@
+     * @param clusterService     Service for cluster state management
+     * @param sslContextProvider Provider for SSL/TLS context configuration
+     * @param elg                Event loop group for network operations
+     * @param threadPool         Thread pool for executing tasks asynchronously
+     * @param client             OpenSearch client
+     */
+    public FlightClientManager(
+        BufferAllocator allocator,
+        ClusterService clusterService,
+        @Nullable SslContextProvider sslContextProvider,
+        EventLoopGroup elg,
+        ThreadPool threadPool,
+        Client client
+    ) {
+        grpcExecutor = threadPool.executor(ServerConfig.FLIGHT_CLIENT_THREAD_POOL_NAME);
+        this.clientConfig = new ClientConfiguration(
+            Objects.requireNonNull(allocator, "BufferAllocator cannot be null"),
+            Objects.requireNonNull(clusterService, "ClusterService cannot be null"),
+            sslContextProvider,
+            Objects.requireNonNull(elg, "EventLoopGroup cannot be null"),
+            Objects.requireNonNull(grpcExecutor, "ExecutorService cannot be null")
+        );
+        this.client = Objects.requireNonNull(client, "Client cannot be null");
+        clusterService.addListener(this);
+    }
+
+    /**
+     * Returns a Flight client for a given node ID.
+     *
+     * @param nodeId The ID of the node for which to retrieve the Flight client
+     * @return An OpenSearchFlightClient instance for the specified node
+     */
+    public FlightClient getFlightClient(String nodeId) {
+        ClientHolder clientHolder = flightClients.getOrDefault(nodeId, null);
+        return clientHolder != null ? clientHolder.flightClient : null;
+    }
+
+    /**
+     * Returns the location of a Flight client for a given node ID.
+     *
+     * @param nodeId The ID of the node for which to retrieve the location
+     * @return The Location of the Flight client for the specified node
+     */
+    public Location getFlightClientLocation(String nodeId) {
+        ClientHolder clientHolder = flightClients.getOrDefault(nodeId, null);
+        return clientHolder != null ? clientHolder.location : null;
+    }
+
+    /**
+     * Builds a client for a given nodeId in asynchronous manner
+     * @param nodeId nodeId of the node to build client for
+     */
+    public void buildClientAsync(String nodeId) {
+        CompletableFuture<Location> locationFuture = new CompletableFuture<>();
+        locationFuture.thenAccept(location -> {
+            DiscoveryNode node = getNodeFromClusterState(nodeId);
+            buildClientAndAddToPool(location, node);
+        }).exceptionally(throwable -> {
+            logger.error("Failed to get Flight server location for node: [{}] {}", nodeId, throwable);
+            throw new RuntimeException(throwable);
+        });
+        requestNodeLocation(nodeId, locationFuture);
+    }
+
+    Map<String, ClientHolder> getClients() {
```

Comment:
You already have `getFlightClients` method that does exactly the same thing

Reply:
removed